New new perldoc parser

jbarrett opened the pull request:
Picking up the work started by @jdorweiler in #209, this parser is intended to capture as much as possible from perldoc.

Some initial changes:

- Uses perldoc-html tarball instead of scraping perldoc.org
- Attempt to pull all docs from index pages
- ~~Use DBI to write output.txt - naming columns seems more robust to me~~

~~New modules required: [DBD::CSV](https://metacpan.org/pod/DBD::CSV), [Text::CSV_XS](https://metacpan.org/pod/Text::CSV_XS)~~

## Parsers

- [X] Command-Line Options (`perl -w`)
- [X] Diagnostics messages (`Allocation too large: %x`)
- [ ] FAQ (`Which version of Perl should I use?`)
- [X] Functions (`map`)
- [X] Glossary (`regex`)
- [X] Operators (`xor operator`)
- [x] POD format codes (`pod L<name>`) ***(~90%)***
- [X] POD commands (`pod =over`)
- [ ] Regular Expression modifiers (`/x regular expression modifier`) ***(~50%)***
- [X] Variables (`perlvar`) (`$ARGV`)

https://duck.co/ia/view/perl_doc

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

jbarrett commented on lib/fathead/perl_doc/Parsers/parse-faq.pl:
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env perl
```
I've been trying to gather all this into parse.pl rather than have a bunch of scripts all doing almost the same thing.

jbarrett replied:
The idea being you can build a hash of indexes, their contents and the required parser and process them fairly uniformly [here](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/6506d2a/lib/fathead/perl_doc/Parsers/parse.pl#L173-L199)

jbarrett replied:
...which may be over complicating things ¯\_(ツ)_/¯

GuiltyDolphin replied:
@jbarrett Hmm, you mean we have a separate routine for any (different) parsing, then a HASH from files (or groups) to the parser to be used to parse the page?

jbarrett replied:
Yeah, basically the name of the parser sub is the value for `$parser` - I'm thinking some parsers will be useful across sets anyway, or some sets will need multiple types of parser. If we stick them in one place, it provides some mix-and-match flexibility.

jbarrett replied:
Just to point out also, this was just my idea of a flexible approach. It may turn out that throwing lists of pages at ad-hoc scripts ends up being simpler and more maintainable. I think the DBI thing is probably the most valuable part of the script - manually counting tab characters doesn't feel robust to me.

Going to leave it in your hands - if you want to keep going with parse-faq, that's absolutely cool :smile:

GuiltyDolphin replied:
@jbarrett No, no - you're idea was good. I've moved the aux parsers into the main file; we should be able to mix n' match parsers and ensure only a minimal amount of functionality is defined uniquely by each parser (easier to have shared routines).

jbarrett commented:
@GuiltyDolphin Are you currently getting output from this (i.e. something I'm missing?)

The indices hash is empty in the last committed version:

```
\ {
    index-faq         undef,
    index-functions   undef,
    index-internals   undef,
    index-language    undef,
    index-modules-A   undef,
    index-modules-B   undef,
    index-modules-C   undef,
    index-modules-D   undef,
    index-modules-E   undef,
    index-modules-F   undef,
    index-modules-G   undef,
    index-modules-H   undef,
    index-modules-I   undef,
    index-modules-L   undef,
    index-modules-M   undef,
    index-modules-N   undef,
    index-modules-O   undef,
    index-modules-P   undef,
    index-modules-S   undef,
    index-modules-T   undef,
    index-modules-U   undef,
    index-modules-X   undef,
    index-platforms   undef,
    index-pragmas     undef,
    index-tutorials   undef,
    index-utilities   undef
}
```
 Thanks!

GuiltyDolphin commented:
@jbarret If you run `./fetch.sh && ./parse.sh` then try with `duckpan query perl_doc` everything should be working fine.

jbarrett commented:
Thanks @GuiltyDolphin! Don't know how I missed your response... Will check it out now!

Seems I was missing the fetch step (which I assumed I didn't need since I had the tarball extracted in the expected place... weird).

Alrighty, think I need to get cracking on perlfunc and perlvar. Thanks!

jbarrett commented:
Covering perlvar now ...
![perlvar](https://cloud.githubusercontent.com/assets/444449/17211629/7d7004c8-54c2-11e6-878f-7ce84436d524.png)
...though the rendering really needs work. My attempt to nuke links is broken (should we remove them or rewrite them to point to perldoc.perl.org?) and code blocks are not rendered.

jbarrett commented:
@GuiltyDolphin I updated the faq parser - seems it was pulling in all `<p>`s remaining in the doc, so you got the answer to your question and every subsequent question.

File size now sits at ~560K.

OK, what's next?! :smile:

GuiltyDolphin commented:
@jbarrett I was thinking we should just have all links point to Perldoc.org - that way people can easily get any additional information.

Whoops, yeah, sorry about that ;) (`<p>`)

I've made a start on the parser for regex pages, I'm hoping we can answer the following queries:

`perl /m modifier`, `perl regex modifiers`, `perl capture groups`, `perl match letter` etc.

jbarrett commented:
:+1: Thanks! As you can see here, I'm trying to capture as mush as possible with the FAQ parser right now - seems to be quite effective!

Let me know if this is stepping on your toes and I can back off.

GuiltyDolphin commented:
@jbarrett Go for it - the FAQ responses are likely to be some of the most useful.

I'm going to play around with the regex parser a bit (and see what else there is); rebasing is going to be fun ;)

jbarrett commented:
This is cool - really starting to take shape! With the right set of aliases, this will be awesome - thanks @GuiltyDolphin!

![perlvar_ s](https://cloud.githubusercontent.com/assets/444449/17249934/17b408ea-559a-11e6-8565-979a17ebbf8f.png)

![perl_oo](https://cloud.githubusercontent.com/assets/444449/17249933/17b43c8e-559a-11e6-87a9-f4e88a694053.png)

GuiltyDolphin commented:
@jbarrett I've done a bunch more parsers - I'm a little worried about how the triggering is going so could we get this deployed to beta for testing?

jbarrett commented:
Ping @jdorweiler ^ What's involved in beta deployment for fathead?

Thanks!

jdorweiler commented:
@jbarrett I'll deploy it to beta for you.

jbarrett commented:
:+1: Thanks @jdorweiler!

How do we define trigger words for this? I assume there are some in place for the previous version of this fathead, just wondering generally.

jdorweiler commented:
@jbarrett How about put the trigger words in a separate file `trigger_words.txt`? I'll have to add them manually to our internal list of triggers but at least they will be documented in a place where everyone can see them.

jbarrett commented:
@GuiltyDolphin Do you think it's worthwhile to change perldoc links to DDG query links (where possible)? I think, for example, our for / foreach result might be more useful when the aliases are in.

jbarrett commented:
@GuiltyDolphin I think FAQs are done, right?

jbarrett commented on lib/fathead/perl_doc/Parsers/parse.pl:
```diff
@@ -739,6 +747,13 @@ sub parse_page {
+sub add_related {
+    my ( $self, $article, @related ) = @_;
+    my $sql = 'UPDATE output.txt SET related = ? WHERE title = ?';
+    my $related = join ",", map { "[[$_]]" } @related;
```
@jdorweiler Hmmm, are related articles comma separated or newline?

Doesn't appear to be in the docs. Thanks!

jbarrett commented:
We can now create custom sets of aliases (and related articles, though I'm not 100% if that's right) to be processed after the docs are parsed.

![screen shot 2016-08-01 at 18 44 25](https://cloud.githubusercontent.com/assets/444449/17303332/26d84c3e-5818-11e6-8949-b786fbdf6ba1.png)

GuiltyDolphin commented on lib/fathead/perl_doc/Parsers/parse.pl:
```diff
@@ -689,6 +689,8 @@ sub normalize_article {
     my ($article) = @_;
     my $text = $article->{text};
     $text =~ s/\n/ /g;
+    $text =~ s/<strong>//g;
```
What's wrong with bold text?

GuiltyDolphin commented:
@jbarrett I'm not sure how much we want to be hardcoding specific aliases (as opposed to forms); I guess it's alright for some *very* popular phrases, but we should probably be sticking to auto-generating them in general.

Yeah, FAQs are pretty much done - I'm not sure how well they'll trigger (we might want to add alias their stop-word-removed versions to themselves); but it should be fine for now :+1:

jbarrett commented:
I think we could use the URL, right? If the link matches the "More at..." for an article we could generate a search for that article's title?

I think we're going to end up creating a lot of ad-hoc aliases (or overrides - the foreach function just needs special handling as far as I can tell. The docs don't link to the most useful article).

GuiltyDolphin commented:
@jbarrett Hmm, yeah, seeing as these will only run once ever' so often I think we could probably get away with a bit more overhead on a 'secondary' parse phase. I'll have a peek.

GuiltyDolphin commented:
@jbarrett Okay, this seems to work pretty well.

Now we create the articles *before* we actually insert them into the DB; then we normalize links (if we can handle the query we link to the search) and insert into the DB.

jbarrett commented:
Cool! I was thinking of iterating over the generated file using the SQL abstraction, but this is likely way faster. Checking it out now - thanks!

jbarrett commented:
I think the code block changes have messed up inline code blocks... should we try to fix them or nix them?

![screen shot 2016-08-02 at 11 07 03](https://cloud.githubusercontent.com/assets/444449/17325056/67b86c82-58a1-11e6-8701-26861b84750b.png)

I'm wondering if there's a way to change links to stuff like 'foreach (function)' and 'for (function)' to point to the aliases we created? These docs genuinely seem useless to me.

GuiltyDolphin commented:
@jbarrett Yeah, inline code has gone a bit messy (but doesn't even show on Beta if not wrapped in `<pre>` tags). I think the best thing to do is just to remove inline `<code>` blocks and just have either normal or bold text?

@jbarrett Hmm, yes, it should be easy enough to create a redirect (with aliases maintained); I'll have a poke.

GuiltyDolphin commented:
@jbarrett Huh... Sometimes the `inline` code seems to work, see for example: https://beta.duckduckgo.com/?q=perl+alias&ia=about

I'm going to add the `inline` code blocks back, and see if I can get the other blocks working.

GuiltyDolphin commented:
@jbarrett Did this just get deployed? The changes don't appear to be showing.

jbarrett commented:
Not that I'm aware of.

GuiltyDolphin commented:
@jbarrett The text is getting 'normalized' in a hard-to-predict manner on Beta:

Locally:

![screenshot-2016-08-02-13 53 17_649_175](https://cloud.githubusercontent.com/assets/8598426/17329007/85125650-58b8-11e6-81ee-636a0eb59a0a.png)

(Plus some examples)

Beta:

![screenshot-2016-08-02-13 54 06_643_224](https://cloud.githubusercontent.com/assets/8598426/17329034/a0a5dc20-58b8-11e6-840a-9532a565246d.png)

(Plus some other stuff)

Are there any full guidelines on this? The docs aren't very informative and things that look fine locally generally get a bit screwed up on beta.

jbarrett commented on lib/fathead/perl_doc/Parsers/parse.pl:
```diff
@@ -829,6 +827,7 @@ sub dom_for_parsing {
     my ($url, $page) = @_;
     my $dom = dom_for_file($page);
     normalize_dom_links($url, $dom);
+    $dom->find('strong')->map('strip');
```
:+1:

GuiltyDolphin commented:
@jbarrett Darn, the related topics don't seem to be working... At all.

It looks like `(perl 5)` is being appended to the names so rather than displaying the `(definition)`, `(function)` etc. part of the title nicely it is just embedding them in the header! I was using these to make a clear distinction between the same words (used in different contexts) - and it looked quite nice.

Local:

![screenshot-2016-08-02-17 07 27_625_151](https://cloud.githubusercontent.com/assets/8598426/17335890/af43e270-58d3-11e6-8f30-2d2888d84010.png)

Beta:

![screenshot-2016-08-02-17 08 22_643_124](https://cloud.githubusercontent.com/assets/8598426/17335904/bf92d99c-58d3-11e6-8363-9130401b5d76.png)

jbarrett commented:
There is a bunch of filtering done to text styling, since I think bolding of words is often quite strange outside the full article context (though don't quote me on that). It may just be for consistency... but yeah, inline bold and code spans get filtered and why / how / under what circumstances it's done does not appear to be sufficiently documented (unless it's in our template docs).

FYI, there's a triggering issue too - I think there's a sanity check somewhere for titles with all non-alpha characters so, e.g. the entry for [$_ is simply not present](https://beta.duckduckgo.com/?q=perl+%24_). FWIW, when we fix this filtering, the triggering should be fine.

It seems we need to do some work on app-duckpan so as not to confound expectations with IA rendering. `duckpan server` is a big smelly second system anyway, so this is roughly in line with what we've been doing to date :smile: 

I've been thinking about the trigger words and linking to related fathead articles (this looks great, btw - nice work!) We'll need to include our tigger word in these links and update app-duckpan to filter them out (I guess a command line option like `--trigger-word="perl"` is good enough?)

GuiltyDolphin commented:
> I've been thinking about the trigger words and linking to related fathead articles (this looks great, btw - nice work!) We'll need to include our tigger word in these links and update app-duckpan to filter them out (I guess a command line option like --trigger-word="perl" is good enough?)

@jbarrett You mean have `foreach (function)` as the display on a related topic, but use `foreach (function) perl` as the link query? Yeah, that might help with the triggering issue.

I'll create a quick PR adding that option - would make it easier to test in an environment closer to beta.

jbarrett commented:
This is looking really great. Apart from the backend filtering / formatting changes we need to do internally, how can I help you get this over the line?

GuiltyDolphin commented:
@jbarrett Well, there's some more parsers I want to add (at some point) - but we can do those later with other PRs.

The output is a bit messy visually on Beta - but I'm not really sure how to combat that without better knowledge of how the normalizing is done.

I'll let you call it when you're happy, here's a few stats w/ the current parsers:

```
Parse Statistics:
Aliases: 5777
Articles: 2662
Disambiguations: 46
Pages: 462
Unique Aliases: 5451
Total entries: 8159
```

jdorweiler commented:
@GuiltyDolphin can you give me some examples of messy results on beta?

GuiltyDolphin commented:
@jdorweiler Has this been deployed recently? I think the results being displayed were from a previous version (some of the formatting may have been fixed).

jdorweiler commented:
@GuiltyDolphin it's up to date now.

jdorweiler commented:
@GuiltyDolphin I think best thing to do is remove the internal links in the code blocks so we can get the new data deployed. We can add them back in later once the problem has been fixed.

GuiltyDolphin commented:
@jdorweiler (Done) - yeah, we can sort out the link issue later (as with the Idris FatHead).

jbarrett commented:
Assigning this to @moollaza and @jdorweiler for review for now. Feel free to punt this on to someone more qualified / available.

@GuiltyDolphin How close do you think we are to ready? I'd like to go live with this soon and see how it affects the coverage figures. We can note further issues in the forum thread?

https://forum.duckduckhack.com/t/improving-the-perldoc-fathead/72

Thanks!

GuiltyDolphin commented:
@jbarrett I think it's close; it's just hard to verify everything as there's so many different formats for the Perl documentation (so any changes *could* break something w/o us noticing) (compare this to the Idris-doc parser where the format is very consistent).

Sure, we can use the forum.

jbarrett commented:
@GuiltyDolphin OK, this is a little messy... I've been using @andrey-p's [ddg-screen-diff](https://github.com/duckduckgo/ddg-screen-diff) tool to pull screenshots of this fathead from beta and batch them into sets of 100. There are a few missing or not triggering here and there, but on the whole it looks very good:

https://john.duckduckgo.com/screens/

(Sorry for the rudimentary presentation)

We're also continuing to work on triggering for non-alpha stuff internally, but as I think I've said elsewhere, that should block us merging this.

jbarrett commented:
@GuiltyDolphin I know you've been checking this out on beta quite a bit, so have spotted quite a few issues we need to fix with the infrastructure.

Other than '$_' and the like not triggering, and the handling of links are there any other issues you've observed?

Many thanks!

GuiltyDolphin commented:
@jbarrett It seems to be mostly triggering and formatting; is the Beta version fully up-to-date?

The definitions (glossary) are looking pretty good, and the diagnostics are mostly correct (apart from triggering).

jbarrett commented:
@GuiltyDolphin I'm a little confused by my output here - are we doing some sort of hex encoding of UTF-8 in parse.pl or is this a weird side effect of something else?

```
℞ 39: Case- accent-insensitive comparisons      A                                                                               <p>Use a collator object to compare Unicode text by character instead of by codepoint.</p><pre class="verbatim"><ol><li> <span class="l_k" href="http://perldoc.perl.org/functions/use.html">use</span> <span class="w">Unicode::Collate</span><span class="sc">;</span></li><li> <span class="l_k" href="http://perldoc.perl.org/functions/my.html">my</span> <span class="i">$es</span> = <span class="w">Unicode::Collate</span><span class="w">-&gt;new</span><span class="s">(</span></li><li>     <span class="w">level</span> <span class="cm">=&gt;</span> <span class="n">1</span><span class="cm">,</span></li><li>     <span class="w">normalization</span> <span class="cm">=&gt;</span> <span class="l_k" href="http://perldoc.perl.org/functions/undef.html">undef</span></li><li> <span class="s">)</span><span class="sc">;</span></li><li></li><li>  <span class="c"># now both are true:</span></li><li> <span class="i">$es</span><span class="i">-&gt;eq</span><span class="s">(</span><span class="q">&quot;Garc\xEDa&quot;</span><span class="cm">,</span>  <span class="q">&quot;GARCIA&quot;</span> <span class="s">)</span><span class="sc">;</span></li><li> <span class="i">$es</span><span class="i">-&gt;eq</span><span class="s">(</span><span class="q">&quot;M\xE1rquez&quot;</span><span class="cm">,</span> <span class="q">&quot;MARQUEZ&quot;</span><span class="s">)</span><span class="sc">;</span></li></ol></pre>  http://perldoc.perl.org/perlunicook.html#%211e-39%3a-Case--_and_-accent-insensitiCannot printf %f with '%c'    A                                                                               <p>(F) You tried printing an infinity or not-a-number as a character (%c), which makes no sense.  Maybe you meant &#39;%s&#39;, or just stringifying it?</p>    http://perldoc.perl.org/perldiag.html#Cannot-printf-%25f-with-'%25c'
```
The original text is:

http://perldoc.perl.org/perlunicook.html#!1e-39%3a-Case--_and_-accent-insensitive-comparisons

García becomes Garc\xEDa
Márquez becomes M\xE1rquez

It also looks like the start of a new entry is stuck on the end of that one. We may have reached the limit of what we can achieve with a DBI abstraction of a TSV? :)

Thanks!

mintsoft commented:
@jbarrett @GuiltyDolphin I think that's an encoding issue. \xE1 is extended ASCII (i.e. totally unstandardised) for á IIRC. Looks like it needs outputting as UTF8 and it'll sort itself out

jbarrett commented:
@mintsoft Weird that we just get that for the abstract and not the title, which contains that weird prescription symbol for some reason. Anyway, it's an avenue to explore... thanks!

As an aside, does the rendering of this character in the [perlunicook content listing](http://perldoc.perl.org/perlunicook.html) look wonked? :smile:

GuiltyDolphin commented:
@jbarrett Yeah, sorry, I was XML-escaping the titles in certain areas; I pushed a commit yesterday which should fix it (hopefully).

GuiltyDolphin commented on lib/fathead/perl_doc/Parsers/parse.pl:
```diff
@@ -286,6 +286,7 @@ sub article {
 sub entry {
     my ( $self, %article ) = @_;
     my ($title, $text, $url, $related) = @article{qw(title text url related)};
+    $text =~ s{\\}{\\\\}g;
```
I'm asking the questions.

jbarrett commented:
@GuiltyDolphin Thanks! I'm working on the output file stuff right now (will be pushed soon, I think  -  basically bypassing the probably-flawed SQL-abstraction approach).

I've deployed the latest commit to beta and updated the screenshots here:

https://john.duckduckgo.com/screens/

jbarrett commented on lib/fathead/perl_doc/Parsers/parse.pl:
```diff
@@ -0,0 +1,952 @@
+        my @to = @$to;
+        @to == 1 and $self->insert_alias($alias, $to[0]) and next;
+        next;
+        my @articles = map { $self->resolve_alias($_) } @to;
```
@GuiltyDolphin I'm a little confused by what's going on from this line to the end of the sub. I've nixed it temporarily since it uses the SQL interface.

It looks like we're creating extra redirects and disambiguation articles from...? Can we use the articles hash for this data?

I guess my main question is, can we live without it for now? Thanks!

GuiltyDolphin replied:
@jbarrett Well, in the IdrisDoc parser I've done it differently, see https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/lib/fathead/idris_doc/parse.pl#L581  and https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/lib/fathead/idris_doc/parse.pl#L211 where I've used the article hash as a lookup.

I think there were some multi-redirects in here as a result of the hard-coded aliases, but apart from that it should be fine to just have what's in the Idris parser.

jbarrett commented:
https://john.duckduckgo.com/screens1/

I am looking for brokenness in **likely** triggered abstracts / articles (in other words, if a FAQ entry or something with a long title which is currently unlikely to be triggered is broken, I don't care :smile:)

I'm also not too concerned about broken triggering since it has not been established whether it's this PR or fathead infrastructure in most cases.

Next I want to restore the for, foreach etc. redirects.

Once they're in place and we've pinned down major breakage (if any) I'd like this to go live.

jbarrett commented:
@GuiltyDolphin I'm running screenshots again right now, they should be published soon. Until the infrastructure is fixed to deal with this data, we need one last piece of double-escape filtering, disambiguation content.

jbarrett commented:
Rendering looks good. Triggering needs some tweaks in cases. I'm happy to call this "RC1" :smile: 

https://john.duckduckgo.com/perldocrc1/

We should probably go live with this as-is and start addressing issues in smaller tasks.

Thanks!

jbarrett commented:
Updated on [beta](https://beta.duckduckgo.com/).

moollaza commented:
@jbarrett overall this seems more the good enough to go live 👍 

I noticed that for https://beta.duckduckgo.com/?q=perl+exists&t=hj&ia=about&iax=1 we have `<br>` before and after `<code>`.

IMO we should avoid `<br>` and wrap any plain text inside `<p>` as both `<p>` and `<pre>` have margins and padding that space out the content nicely. The `<br>` creates a little too much whitespace.

moollaza commented:
I see now that we're not wrapping the initial `<code>` inside a ` <pre>` -- I believe the expectation of our CSS is that any `<code>` lives inside a `<pre>`

GuiltyDolphin commented:
@moollaza There are a few formatting issues (mostly minor) that should be fixed in the future - perhaps we should open an issue (as a reminder) and fix this at a later date?

@moollaza Any `code` in a `pre` is displayed literally.

jdorweiler commented:
Looks good to me. We can just make issues for the smaller stuff.

moollaza commented on lib/fathead/perl_doc/Parsers/parse.pl:
```diff
@@ -0,0 +1,963 @@
+        uls => \@mod_sections,
+        title => sub { $_[0]->find('b')->first->text . ' (variable)' },
+        aliases => sub { $_[1] =~ s/ \(variable\)//r },
+        categories => sub { ['Variables'] },
```
These need to be more specific as categories are created globally. i.e "Perl Variables"

Otherwise the search "variables" would show this category, which isn't expected

jbarrett replied:
@moollaza Interesting. Do you have a specific example of where this is an issue?

Thanks!

GuiltyDolphin replied:
@moollaza Whoops, yeah - we should probably sort out categories entirely while we're at it.
